README.md
# overlay_notification

[![Pub](https://img.shields.io/pub/v/overlay_notification?color=blue&style=plastic)](https://pub.dev/packages/overlay_notification)
[![CI](https://github.com/mohesu/overlay_notification/workflows/CI/badge.svg?style=plastic)](https://github.com/mohesu/overlay_notification/actions)

Provider support for `overlay`, make it easy to build **toast** and **In-App notification**.

**this library support ALL platform**

## Interaction

If you want to see the ui effect of this library, just click
here [https://mohesu.github.io/overlay_notification/#/](https://mohesu.github.io/overlay_notification/#/)

## How To Use

#### Add dependencies into you project `pubspec.yaml` file

```yaml
    dependencies:
        overlay_notification: latest_version
```
the latest version is [![Pub](https://img.shields.io/pub/v/overlay_notification?color=blue&style=plastic)](https://pub.dev/packages/overlay_notification)

#### Wrap your AppWidget with `OverlayNotification`

```dart #build()
  return OverlayNotification.global(child: MaterialApp());
```

#### Show toast or simple notifications

```dart
showSimpleNotification(
     Text("this is a message from simple notification"),
     background: Colors.green,
);

toast('this is a message from toast');

```

#### Notification with decoration

```dart
showSimpleNotification(
  const Text('this is a message from simple notification with decoration'),
    margin: const EdgeInsets.all(38),
    decoration: BoxDecoration(
      borderRadius: BorderRadius.circular(18),
      gradient: const LinearGradient(
      colors: [Colors.green, Colors.blue],
      begin: Alignment.topLeft,
      end: Alignment.bottomRight),
     ),
    elevation: 0,
    background: Colors.transparent,
);
```

#### Popup a custom notification

you can custom your notification widget to popup, for example:

```dart
showOverlayNotification((context) {
  return MessageNotification(
    message: 'i love you',
    onReplay: () {
      OverlayNotificationEntry.of(context).dismiss(); //use OverlaySupportEntry to dismiss overlay
      toast('you checked this message');
    },
  );
```

```dart MessageNotification Class
class MessageNotification extends StatelessWidget {
  final VoidCallback onReplay;

  const MessageNotification({Key key, this.onReplay}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Card(
      margin: const EdgeInsets.symmetric(horizontal: 4),
      child: SafeArea(
        child: ListTile(
          leading: SizedBox.fromSize(
              size: const Size(40, 40),
              child: ClipOval(child: Image.asset('assets/avatar.png'))),
          title: Text('Lily MacDonald'),
          subtitle: Text('Do you want to see a movie?'),
          trailing: IconButton(
              icon: Icon(Icons.reply),
              onPressed: () {
                ///TODO i'm not sure it should be use this widget' BuildContext to create a Dialog
                ///maybe i will give the answer in the future
                if (onReplay != null) onReplay();
              }),
        ),
      ),
    );
  }
}
```

#### Popup a custom overlay

for example, we need create a iOS Style Toast.

```dart
  showOverlay((context, t) {
    return Opacity(
      opacity: t,
      child: IosStyleToast(),
    );
  });
```

```dart
class IosStyleToast extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return SafeArea(
      child: DefaultTextStyle(
        style: Theme.of(context).textTheme.body1.copyWith(color: Colors.white),
        child: Padding(
          padding: const EdgeInsets.all(16),
          child: Center(
            child: ClipRRect(
              borderRadius: BorderRadius.circular(10),
              child: Container(
                color: Colors.black87,
                padding: const EdgeInsets.symmetric(
                  vertical: 8,
                  horizontal: 16,
                ),
                child: Column(
                  mainAxisSize: MainAxisSize.min,
                  children: <Widget>[
                    Icon(
                      Icons.check,
                      color: Colors.white,
                    ),
                    Text('Succeed')
                  ],
                ),
              ),
            ),
          ),
        ),
      ),
    );
  }
}

```

## End

if you have some suggestion or advice, please open an issue to let me known.
This will greatly help the improvement of the usability of this project.
Thanks.

## Thanks to

`overlay_support` https://pub.dev/packages/overlay_support

This package is inspired by `overlay_support`, but it is not compatible with `overlay_support` and `overlay_support` is not compatible with `overlay_notification`.


CHANGELOG.md
## 0.0.2
- test added

## 0.0.1
- initial release


example/lib/pages/page_main.dart
import 'dart:async';
import 'dart:math';

import 'package:example/notification/custom_animation.dart';
import 'package:example/notification/custom_notification.dart';
import 'package:example/notification/ios_toast.dart';
import 'package:flutter/material.dart';
import 'package:overlay_notification/overlay_notification.dart';

class HomePage extends StatelessWidget {
  const HomePage({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return ListView(
      children: <Widget>[
        _Section(title: 'Notification', children: <Widget>[
          ElevatedButton(
            onPressed: () {
              showSimpleNotification(
                const Text(
                    'this is a message from simple notification with decoration'),
                margin: const EdgeInsets.all(38),
                decoration: BoxDecoration(
                  borderRadius: BorderRadius.circular(18),
                  gradient: const LinearGradient(
                    colors: [Colors.green, Colors.blue],
                    begin: Alignment.topLeft,
                    end: Alignment.bottomRight,
                  ),
                ),
                elevation: 0,
                background: Colors.transparent,
              );
            },
            child: const Text(
              'Auto Dismiss Notification',
            ),
          ),
          ElevatedButton(
            onPressed: () {
              showSimpleNotification(
                const Text('you got a simple message'),
                trailing: Builder(builder: (context) {
                  return TextButton(
                      onPressed: () {
                        OverlayNotificationEntry.of(context)!.dismiss();
                      },
                      child: const Text('Dismiss',
                          style: TextStyle(color: Colors.amber)));
                }),
                background: Colors.green,
                autoDismiss: false,
                slideDismissDirection: DismissDirection.up,
              );
            },
            child: const Text('Fixed Notification'),
          ),
          ElevatedButton(
            child: const Text('Bottom Notification'),
            onPressed: () {
              showSimpleNotification(
                const Text('Hello'),
                position: NotificationPosition.bottom,
                slideDismissDirection: DismissDirection.down,
              );
            },
          )
        ]),
        _Section(title: 'Custom notification', children: <Widget>[
          ElevatedButton(
            onPressed: () {
              showOverlayNotification((context) {
                return MessageNotification(
                  message: messages[3],
                  onReply: () {
                    OverlayNotificationEntry.of(context)!.dismiss();
                    toast('you checked this message');
                  },
                );
              }, duration: const Duration(milliseconds: 4000));
            },
            child: const Text('custom message notification'),
          ),
          ElevatedButton(
            onPressed: () async {
              final random = Random();
              for (var i = 0; i < messages.length; i++) {
                await Future.delayed(
                    Duration(milliseconds: 200 + random.nextInt(200)));
                showOverlayNotification((context) {
                  return MessageNotification(
                    message: messages[i],
                    onReply: () {
                      OverlayNotificationEntry.of(context)?.dismiss();
                      toast('you checked this message');
                    },
                  );
                },
                    duration: const Duration(milliseconds: 4000),
                    key: const ValueKey('message'));
              }
            },
            child: const Text('message sequence'),
          ),
        ]),
        _Section(title: 'toast', children: [
          ElevatedButton(
            onPressed: () {
              toast('this is a message from toast');
            },
            child: const Text('toast'),
          )
        ]),
        _Section(title: 'custom', children: [
          ElevatedButton(
            onPressed: () {
              showOverlay((_, t) {
                return Theme(
                  data: Theme.of(context),
                  child: Opacity(
                    opacity: t,
                    child: const IosStyleToast(),
                  ),
                );
              }, key: const ValueKey('hello'));
            },
            child: const Text('show iOS Style Dialog'),
          ),
          ElevatedButton(
            onPressed: () {
              showOverlay((context, t) {
                return CustomAnimationToast(value: t);
              }, key: const ValueKey('hello'), curve: Curves.decelerate);
            },
            child: const Text('show custom animation overlay'),
          ),
          ElevatedButton(
            onPressed: () {
              showOverlay((context, t) {
                return Container(
                  color: Color.lerp(Colors.transparent, Colors.black54, t),
                  child: FractionalTranslation(
                    translation: Offset.lerp(
                        const Offset(0, -1), const Offset(0, 0), t)!,
                    child: Column(
                      children: <Widget>[
                        MessageNotification(
                          message: 'Hello',
                          onReply: () {
                            OverlayNotificationEntry.of(context)!.dismiss();
                          },
                          key: const ModalKey(Object()),
                        ),
                      ],
                    ),
                  ),
                );
              }, duration: Duration.zero);
            },
            child: const Text('show notification with barrier'),
          )
        ])
      ],
    );
  }
}

class _Section extends StatelessWidget {
  final String title;

  final List<Widget> children;

  const _Section({Key? key, required this.title, required this.children})
      : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.symmetric(horizontal: 8),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: <Widget>[
          const SizedBox(height: 16),
          _Title(title: title),
          Wrap(spacing: 8, children: children),
        ],
      ),
    );
  }
}

class _Title extends StatelessWidget {
  final String title;

  const _Title({Key? key, required this.title}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Container(
      margin: const EdgeInsets.only(left: 16, top: 10, bottom: 8),
      child: Text(
        title,
        style: const TextStyle(fontWeight: FontWeight.bold, fontSize: 17),
      ),
    );
  }
}
